src/topp/MultiplexResolver.cpp

```diff
@@ -0,0 +1,771 @@
+/// @cond TOPPCLASSES
+
+class TOPPMultiplexResolver :
+  public TOPPBase
+{
+private:
+
+  // input and output files
+  String in_;
+  String out_;
+  String out_conflicts_;
+
+  // section "algorithm"
+  String labels_;
+  std::vector<std::vector<String> > samples_labels_;
+  unsigned charge_min_;
+  unsigned charge_max_;
+  unsigned missed_cleavages_;
+  unsigned isotopes_per_peptide_min_;
+  unsigned isotopes_per_peptide_max_;
+  double mass_tolerance_;
+
+  // section "labels"
+  map<String, double> label_mass_shift_;
+  
+public:
+  TOPPMultiplexResolver() :
+    TOPPBase("MultiplexResolver", "Completes peptide multiplets and resolves conflicts within them.", true),
+    charge_min_(1), charge_max_(1), missed_cleavages_(0), isotopes_per_peptide_min_(1), isotopes_per_peptide_max_(1), mass_tolerance_(0.1)
+  {
+  }
+  
+  void registerOptionsAndFlags_()
+  {
+    registerInputFile_("in", "<file>", "", "Peptide multiplets with assigned sequence information");
+    setValidFormats_("in", ListUtils::create<String>("consensusXML"));
+    registerOutputFile_("out", "<file>", "", "Complete peptide multiplets.", false);
+    setValidFormats_("out", ListUtils::create<String>("consensusXML"));
+    registerOutputFile_("out_conflicts", "<file>", "", "Optional output containing peptide multiplets with conflicting quant/ID information.", false, true);
+    setValidFormats_("out_conflicts", ListUtils::create<String>("consensusXML"));
+
+    registerSubsection_("algorithm", "Parameters for the algorithm.");
+    registerSubsection_("labels", "Isotopic labels that can be specified in section \'algorithm:labels\'.");
+
+  }
+
+  // create parameters for sections (set default values and restrictions)
+  Param getSubsectionDefaults_(const String& section) const
+  {
+    Param defaults;
+
+    if (section == "algorithm")
+    {
+      defaults.setValue("labels", "[][Lys8,Arg10]", "Labels used for labelling the samples. [...] specifies the labels for a single sample. For example\n\n[][Lys8,Arg10]        ... SILAC\n[][Lys4,Arg6][Lys8,Arg10]        ... triple-SILAC\n[Dimethyl0][Dimethyl6]        ... Dimethyl\n[Dimethyl0][Dimethyl4][Dimethyl8]        ... triple Dimethyl\n[ICPL0][ICPL4][ICPL6][ICPL10]        ... ICPL");
+      defaults.setValue("charge", "1:4", "Range of charge states in the sample, i.e. min charge : max charge.");
+      defaults.setValue("isotopes_per_peptide", "3:6", "Range of isotopes per peptide in the sample. For example 3:6, if isotopic peptide patterns in the sample consist of either three, four, five or six isotopic peaks. ", ListUtils::create<String>("advanced"));
+      defaults.setValue("missed_cleavages", 0, "Maximum number of missed cleavages due to incomplete digestion.");
+      defaults.setMinInt("missed_cleavages", 0);
+      defaults.setValue("mass_tolerance", 0.1, "Mass tolerance for matching the detected to the theoretical mass shifts.", ListUtils::create<String>("advanced"));
```

Comment:
Say, a delta mass of 8.01 Da was expected but a shift of 8.05 Da was detected. The tolerance determines how far both can deviate.